### PR TITLE
Move timeout, affinity, nodeSelector to taskrun and pipelinerun

### PR DESCRIPTION
**Pull request by pivotal-nader-ziada**
## Proposed changes
- move timeout, affinity, nodeselector from task to taskrun
- move timeout, affinity, nodeselector from pipeline to pipelinerun

These are runtime parameters that are more concerned with the actual
running of a task more than the definition

/cc @bobcatfish @shashwathi

### CODE REVIEW

**Comment by bobcatfish** [comment]
o no my docs

**Comment by pivotal-nader-ziada** [comment]
> o no my docs

This will rebase on your docs once its merged in and I'll update the docs accordingly

**Comment by abayer** [comment]
I'm not convinced that `Timeout` should be on `PipelineRun` and `TaskRun` rather than `Pipeline` and `Task` - for example, this would mean that there's no way to have a single `TaskRun` created by a `PipelineRun` to have its own `Timeout`. In Jenkins Pipeline, there's heavy usage of timeouts for specific stages (i.e., `Task` equivalents) and whole pipelines as part of authoring the pipeline, and I think we really need the same behavior here.

**Comment by pivotal-nader-ziada** [comment]
@abayer this is a valid point, I was just proposing this change because it doesn't seem like there is agreement on where timeout should live. 
I thought about having timeout in the `run` because the task and pipeline are just like templates for what steps are needed and everything else about the actual running of these steps comes from the `run`, for example the `service account` and values of params and so on

**Comment by abayer** [comment]
@pivotal-nader-ziada - I guess I think of `timeout` as part of the "template" in that case - though being able to override the `timeout` in a `Task` or `Pipeline` via `TaskRun`/`PipelineRun` sounds like a good idea to me.

File: pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go [hunk 1]
```diff
@@ -337,19 +337,22 @@ func updateTaskRunsStatus(pr *v1alpha1.PipelineRun, pipelineState []*resources.R
...
-func (c *Reconciler) createTaskRun(logger *zap.SugaredLogger, rprt *resources.ResolvedPipelineRunTask, pr *v1alpha1.PipelineRun, sa, storageBasePath string, pipelineTimeout *metav1.Duration) (*v1alpha1.TaskRun, error) {
-	ts := rprt.ResolvedTaskResources.TaskSpec.DeepCopy()
-	if pipelineTimeout != nil {
-		pTimeoutTime := pr.Status.StartTime.Add(pipelineTimeout.Duration)
-		if ts.Timeout == nil || time.Now().Add(ts.Timeout.Duration).After(pTimeoutTime) {
+func (c *Reconciler) createTaskRun(logger *zap.SugaredLogger, rprt *resources.ResolvedPipelineRunTask, pr *v1alpha1.PipelineRun, sa, storageBasePath string) (*v1alpha1.TaskRun, error) {
+	var taskRunTimeout = &metav1.Duration{Duration: 0 * time.Second}
+	if pr.Spec.Timeout != nil {
+		pTimeoutTime := pr.Status.StartTime.Add(pr.Spec.Timeout.Duration)
+		if time.Now().After(pTimeoutTime) {
```
Comment (shashwathi) [inline pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go]: Can we simplify the timeout logic?  I completely agree change is done in another PR. 

```
taskRunTimeout = &metav1.Duration{Duration: 0 * time.Second}
if pr.Spec.Timeout != nil {
    pTimeoutTime := pr.Status.StartTime.Add(pr.Spec.Timeout.Duration)
              // set taskrun timeout only if its valid timeout
          if !time.Now().After(pTimeoutTime) {
            taskRunTimeout = pr.Spec.Timeout
        }
    } else {
     // set timeout as nil if PR timeout is not set
        taskRunTimeout = nil
    }
```

**Comment by shashwathi** [comment]
I am adding hold to merge @bobcatfish docs PR first and address @abayer comments. This PR is good from my end

**Comment by abayer** [comment]
Actually, I'm going to hold off on that for the moment - lemme see how what you've implemented works out in practice. I still feel that there's a need to be able to say "if the execution of this `Task` in this `Pipeline` takes longer than X, time out", but I may be wrong.

**Comment by bobcatfish** [comment]
> Actually, I'm going to hold off on that for the moment - lemme see how what you've implemented works out in practice. I still feel that there's a need to be able to say "if the execution of this `Task` in this `Pipeline` takes longer than X, time out", but I may be wrong.

Sounds good! We can definitely iterate on this @abayer :D
